client/packages/common/src/hooks/useElectronClient/useElectronClient.ts

```diff
@@ -0,0 +1,101 @@
+import { useState, useEffect } from 'react';
+
+const DISCOVERY_TIMEOUT = 5000;
+
+// Should match server/server/src/discovery.rs (FrontEndHost)
+export type FrontEndHost = {
+  id: string;
+  protocol: 'http' | 'https';
+  port: number;
+  ip: string;
+  name: string;
+  clientVersion: string;
+  // This one is set by electron.ts
+  isLocal: boolean;
+};
+
+type ServerDiscovered = (event: object, value: FrontEndHost) => void;
+export type ElectronAPI = {
+  // When in discovery mode (initiated by startServerDiscovery) this event is emitted when server is discovered (existing or new)
+  serverDiscovered: (eventInfo: ServerDiscovered) => void;
+  // Starts server discovery (connectToServer stops server discovery)
+  startServerDiscovery: () => void;
+  // Asks client to connect to server (causing window to navigate to server url and stops discovery)
+  connectToServer: (server: FrontEndHost) => void;
+  // Will return currently connected client (to display in UI)
+  connectedServer: () => Promise<FrontEndHost | null>;
+};
+
+declare global {
+  interface Window {
+    electronAPI: ElectronAPI;
+  }
+}
+
+type ElectronClientState = {
+  servers: FrontEndHost[];
```

Comment:
changed to an array - the elements in the object were unique, but the object itself was being recreated every time a server was found, which it does every few seconds. It meant that there were more renders than necessary and it was hard to use it as a reference.

Also - the usage was always converting to an array using `Object` - having an array in state makes it easier to consume.

Reply:
I started with Array but misused setState and though object would be easier. Array much cleaner ; ) thanks